st2client/st2client/models/core.py

```diff
@@ -879,3 +879,37 @@ def list(self, group_id, **kwargs):
             result.append(item)
 
         return result
+
+class KeyValuePairResourceManager(ResourceManager):
+    @add_auth_token_to_kwargs_from_env
+    def get_by_name(self, name, **kwargs):
+
+        token = kwargs.get("token", None)
+        api_key = kwargs.get("api_key", None)
+        params = kwargs.get("params", {})
+
+        for k, v in six.iteritems(kwargs):
+            # Note: That's a special case to support api_key and token kwargs
+            if k not in ["token", "api_key", "params"]:
+                params[k] = v
+
+        url = "/%s/%s/?%s" % (
+            self.resource.get_url_path_name(),
+            name,
+            urllib.parse.urlencode(params),
+        )
```

Comment:
So this `get_by_name` is essentially a copy of `_query_details` with but it adds name in the path, right?

I wonder if there's a way to extend `_query_details` instead of duplicating all of this auth logic. maybe we could add a special dunder arg like `__path__` that gets added right after the url. So, in `_query_params()`, this would look like:

```python
for k, v in six.iteritems(kwargs):
    # ...
    if k not in ["token", "api_key", "params", "__path__"]:
        params[k] = v

url = "/%s/%s?%s" % (
    self.resource.get_url_path_name(),
    kwargs.get("__path__", ""),
    urllib.parse.urlencode(params),
)
```

So that this method would call 
```python
instances, _ = _query_details(__path__=f"{name}/", **kwargs)
return instances[0]
```

Reply:
Fair enough. As far as  I understand just extending _query_details is not enough, we need to extend `query` function as well. Please let me know if it makes sense and if I should extend both `query`, `_query_details` functions or should I go ahead with suggested changes(extending `get_by_name` function).

Reply:
Let's keep the new `KeyValuePairResourceManager`, but simplify the implementation of the get by name method. That method does not need to use `query`, it can just directly use `_query_details`.

Reply:
The problem with using _query_details is that it expects multiple results and calls deserialise on each item. In our case '/v1/keys/{keyname}' is called and returns a single object.  This following line fails in our case
```
instances = [self.resource.deserialize(item) for item in items]
```

For this to work we need change _query_details so that it can handle both single, multiple objects. That's the reason I extended get_by_name directly and changed it. Do you want me to extend _query_details and change it such that it can handle single, multiple results or go ahead with my original implementation?

I personally wouldn’t want to modify query_details as this is used by a lot of functions and I don’t know the codebase enough to make these changes confidently without making sure it won’t have any adverse affect

Reply:
Bummer. I guess the code duplication is fine in this case.